lib/resources/tasks/watch.ts

```diff
@@ -0,0 +1,100 @@
+import * as gulp from 'gulp';
+import * as minimatch from "minimatch";
+import * as gulpWatch from "gulp-watch";
+import * as debounce from "debounce";
+import { build } from 'aurelia-cli';
+import * as project from '../aurelia.json';
+import transpile from './transpile';
+import processMarkup from './process-markup';
+import processCSS from './process-css';
+import copyFiles from './copy-files';
+
+const debounceWaitTime = 100;
+let isBuilding = false;
+let pendingRefreshPaths = [];
+let watches = {};
+let watchCallback = () => { };
+
+watches[project.transpiler.source] = { name: "transpile", callback: transpile };
+watches[project.markupProcessor.source] = { name: "markup", callback: processMarkup };
+watches[project.cssProcessor.source] = { name: "CSS", callback: processCSS };
+if (typeof project.build.copyFiles === 'object') {
+  for (let src of Object.keys(project.build.copyFiles)) {
+    watches[src] = { name: "file copy", callback: copyFiles };
+  }
+}
+
+let watch = (callback?) => {
+  watchCallback = callback || watchCallback;
+  return gulpWatch(
+    Object.keys(watches),
+    {
+      read: false, // performance optimization: do not read actual file contents
+      verbose: true
+    },
+    (vinyl) => {
+      if (vinyl.path && vinyl.cwd && vinyl.path.startsWith(vinyl.cwd)) {
+        let pathToAdd = vinyl.path.substr(vinyl.cwd.length + 1);
+        log(`Watcher: Adding path ${pathToAdd} to pending build changes...`);
+        pendingRefreshPaths.push(pathToAdd);
+        refresh();
+      }
+    });
+};
+
+let refresh = debounce(() => {
+  if (isBuilding) {
+    log("Watcher: A build is already in progress, deferring change detection...");
+    return;
+  }
+
+  isBuilding = true;
+
+  let paths = pendingRefreshPaths.splice(0);
+  let refreshTasks = [];
```

Comment:
Well, it is the paths (files) pending to be refreshed and it is the tasks that's going to refresh them so I don't really think so. If you feel the names aren't clear enough I could try coming up with some better ones.